function/transform/special.go

```diff
@@ -157,3 +157,90 @@ var Alias = function.MetricFunction{
 		return list, nil
 	},
 }
+
+// Derivative is special because it needs to get one extra data point to the left
+// This transform estimates the "change per second" between the two samples (scaled consecutive difference)
+var Derivative = newDerivativeBasedTransform("derivative", derivative)
+
+func derivative(values []float64, parameters []function.Value, scale float64) ([]float64, error) {
+	result := make([]float64, len(values)-1)
+	for i := range values {
+		if i == 0 {
+			continue
+		}
+		// Scaled difference
+		result[i-1] = (values[i] - values[i-1]) / scale
+	}
+	return result, nil
+}
+
+// Rate is special because it needs to get one extra data point to the left
+// This transform functions exactly like Derivative but bounds the result to be positive.
+// That is, it returns consecutive differences which are at least 0.
+var Rate = newDerivativeBasedTransform("rate", rate)
+
+func rate(values []float64, parameters []function.Value, scale float64) ([]float64, error) {
+	result := make([]float64, len(values)-1)
+	for i := range values {
+		if i == 0 {
+			continue
+		}
+		// Scaled difference
+		result[i-1] = (values[i] - values[i-1]) / scale
+		if result[i-1] < 0 {
+			// values[i] is our best approximatation of the delta between i-1 and i
+			result[i-1] = math.Max(values[i], 0) / scale
+		}
+	}
+	return result, nil
+}
+
+// newDerivativeBasedTransform returns a function.MetricFunction that performs
+// a delta between two data points. The transform parameter is a function of type
+// transform is expected to return an array of values whose length is 1 less
+// than the given series
+func newDerivativeBasedTransform(name string, transformer transform) function.MetricFunction {
+	return function.MetricFunction{
+		Name:         "transform." + name,
+		MinArguments: 1,
+		MaxArguments: 1,
+		Compute: func(context function.EvaluationContext, arguments []function.Expression, groups function.Groups) (function.Value, error) {
+			var err error
+			// Calcuate the new timerange to include one extra point to the left
+			newContext := context
+			timerange := context.Timerange
+			newContext.Timerange, err = api.NewSnappedTimerange(timerange.Start()-timerange.ResolutionMillis(), timerange.End(), timerange.ResolutionMillis())
```

Comment:
`err` isn't being checked.

An error is very unlikely, but it would occur for timeranges obscenely far in the past, such that the Start time wraps around. The resulting timerange in that case would be invalid, which could cause panics down the line due to division by zero.

If we just don't want to worry about this case, then it would be better to explicitly ignore the error with `_`, and then you don't need to define `var err error` at the top.